### PR TITLE
NAS-113713 / Bluefin / NAS-113713: Refactoring Snapshot List

### PR DESCRIPTION
**Pull request by denysbutenko**
PR is up for early feedback. It is greatly appreciated.

Go to `/storage/snapshots` to check it.

What's new?

Updated components:
- [x] refactor `snapshot-list` to use ix-tables
- [x] add unit tests for `snapshot-list`
- [x] add `snapshot-rollback-dialog` component
- [x] add unit tests for `snapshot-rollback-dialog`
- [x] add `snapshot-batch-delete-dialog` component
- [x] add unit tests for `snapshot-batch-delete-dialog`
- [x] refactor `snapshot-add` component to use ix-forms
- [x] add unit tests for `snapshot-add-form`

Changes:

- [x] add `toolbar-slide-toggle` component  it's for `Entity Toolbar` (use it on Users page)
- [x] improve `slide-toggle` behavior when user clicks `Cancel`
- [x] lazy loading of `SnapshotsModule`
- [x] use store for snapshots
- [x] use store for `showExtraColumns`, previously `snapshotXtraCols` in local storage
- [x] add loading indicator on toggle column (Users page should adapt these improvements)
- [x] remove `SnapshotListRow` interface, using `ZfsSnapshot` instead
- [x] Address feedback items
- [x] fix unit tests

Legacy:
- [x] ~refactor `snapshot-details` to use `ix-expand-toggle-column`~ (removed)
- [x] ~add pipe `convertBytestoHumanReadable`~ (replaced with filesize)
- [x] ~add unit tests for `convertBytestoHumanReadable`~ (replaced with filesize)

### CODE REVIEW

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+  ngOnInit(): void {
+    this.setupToolbar();
+    this.getSnapshots();
+  }
+
+  ngAfterViewInit(): void {
+    this.store$.dispatch(loadSnapshots({ extra: this.showExtraColumns$.value }));
+    this.modalService.onClose$.pipe(untilDestroyed(this)).subscribe(() => {
+      this.getSnapshots();
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts:107]: You are both manually reloading here and listening to updates in effects. Seems like one of the approaches should be used.

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+              used: parseInt(row.properties.used.rawvalue),
+              created: row.properties.creation.parsed.$date,
+              referenced: parseInt(row.properties.referenced.rawvalue),
+            });
+          } else {
+            snapshotListRows.push(transformedRow);
+          }
+        }
+
+        return snapshotListRows;
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts:139]: This processing belongs to a selector (or alternatively you can avoid transforming at all and just read necessary values in a template).

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+  isLoading$ = this.store$.select(selectSnapshotState).pipe(map((state) => state.isLoading));
+  isEmpty$ = this.store$.select(selectSnapshotsTotal).pipe(map((total) => total === 0));
+  @ViewChild(MatSort, { static: false }) sort: MatSort;
+  dataSource: MatTableDataSource<SnapshotListRow> = new MatTableDataSource([]);
+  defaultSort: Sort = { active: 'snapshot', direction: 'desc' };
+  filterString = '';
+  expandedRow: SnapshotListRow;
+  selection = new SelectionModel(true, []);
+  settingsEvent$: Subject<CoreEvent> = new Subject();
+  showExtraColumns$ = new BehaviorSubject(false);
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts:58]: 1. This setting belongs to a state. You should be reading it via selectors and updating it by dispatching actions.
2. This setting better belongs to user preferences, not local storage. Pull and see preferences state slice.

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    this.dataSource.sort = this.sort;
+  }
+
+  expandRow(row: SnapshotListRow): void {
+    this.expandedRow = this.expandedRow === row ? null : row;
+    this.cdr.markForCheck();
+  }
+
+  toggleExtraColumnsDialog(): void {
+    this.showExtraColumns$.next(!this.showExtraColumns$.value);
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts:180]: This should dispatch showExtraColumnsPressed() and this should do processing / trigger the reload.

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+      data: snapshot.id,
+    });
+  }
+
+  doRollback(row: SnapshotListRow): void {
+    this.loader.open();
+    this.ws.call('zfs.snapshot.query', [[['id', '=', row.name]]]).pipe(
+      map((snapshots) => snapshots[0]),
+      tap((snapshot) => {
+        snapshotChanged({ snapshot });
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts:282]: This action doesn't do anything.

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+    this.modalService.openInSlideIn(SnapshotAddComponent);
+  }
+
+  doClone(snapshot: SnapshotListRow): void {
+    this.matDialog.open(SnapshotCloneDialogComponent, {
+      data: snapshot.id,
+    });
+  }
+
+  doRollback(row: SnapshotListRow): void {
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts:277]: You can move the logic of querying for additional snapshot info to SnapshotRollbackDialog itself.

File: src/app/pages/storage/snapshots/store/snapshot.actions.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { createAction, props } from '@ngrx/store';
+import { ZfsSnapshot } from 'app/interfaces/zfs-snapshot.interface';
+
+export const loadSnapshots = createAction('[Snapshots API] Load', props<{ extra: boolean }>());
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/store/snapshot.actions.ts:4]: Prefer actions as events, not commands. In this case, this should be snapshotPageEntered or something like this. Extra argument belongs elsewhere.

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+    }
+  }
+
+  setupToolbar(): void {
+    this.settingsEvent$.pipe(
+      untilDestroyed(this),
+    ).subscribe((event: CoreEvent) => {
+      switch (event.data.event_control) {
+        case 'filter':
+          this.filterString = event.data.filter;
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts:212]: This is probably not correct from state management perspective. ngrx state is supposed to be the source of truth, here we have two sources of truth: MatDataSource and the state. I don't have a good answer on what to do with it at the moment, but we'd need to figure this out before refactoring another table.

File: src/app/pages/storage/snapshots/snapshot-rollback-dialog/snapshot-rollback-dialog.component.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      force: true,
+    };
+    if (this.form.value.recursive) {
+      body.recursive = true;
+    }
+    this.loader.open();
+    console.info('onSubmit', [this.snapshot.name, body]);
+
+    this.ws.call('zfs.snapshot.rollback', [this.snapshot.name, body]).pipe(
+      untilDestroyed(this),
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-rollback-dialog/snapshot-rollback-dialog.component.ts:87]: Did you test this? :)

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.html [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+      <th mat-header-cell *matHeaderCellDef></th>
+      <td mat-cell *matCellDef="let snapshot; dataSource: dataSource">
+        <button
+          [title]="'Settings' | translate"
+          mat-icon-button
+          [matMenuTriggerFor]="actionsMenu"
+        >
+          <mat-icon>more_vert</mat-icon>
+        </button>
+        <mat-menu #actionsMenu="matMenu">
```
Comment (undsoft) [inline src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.html:80]: An argument can be made that behaiour should be consistent between two views as not to be unexpected to the user.

**Review by undsoft — CHANGES_REQUESTED** [review]
Looks pretty good.

Some issues that came up during testing:
1. Initial page loading hard-freezes my browser for 10~30 seconds with 600 snapshots and no extra columns. I understand that we need to do some processing, but 600 is not that much. See if it can be improved in some way. 
2. When one snapshot is selected, it says "All 1 snapshots on this page are selected". "All" and "on this page" are confusing, since I can select entries across different pages.
3. Styles for Select all snapshots do not match the rest of the system. Also, I'm not really sure we need this button. There are very few scenarios where I can see a user selecting all of the snapshots.
4. Expandable row is rendered in DOM when collapsed. Not good for performance and RE team will not be happy.

**Comment by denysbutenko** [comment]
> Looks pretty good.
> 
> Some issues that came up during testing:
> 
> 1. Initial page loading hard-freezes my browser for 10~30 seconds with 600 snapshots and no extra columns. I understand that we need to do some processing, but 600 is not that much. See if it can be improved in some way.
> 2. When one snapshot is selected, it says "All 1 snapshots on this page are selected". "All" and "on this page" are confusing, since I can select entries across different pages.
> 3. Styles for Select all snapshots do not match the rest of the system. Also, I'm not really sure we need this button. There are very few scenarios where I can see a user selecting all of the snapshots.
> 4. Expandable row is rendered in DOM when collapsed. Not good for performance and RE team will not be happy.

1. Yeah, definitely need to find a way. I've tested it on local truenas instance with 1k+ snapshots while refactoring. It works well but sometimes I get freezes as well. Probably we can achieve better performance if add backend pagination with state management. Did you try to disable the Store Devtools extension as a workaround?
2 and 3 ok, let's skip it
4. yeah, I'll create another ticket for research. In unit tests, I added a method for TableHarness to skip expanded rows as a temp workaround.

Thanks for the feedback  👍🏻  I'll address it soon

**Comment by undsoft** [comment]
> Probably we can achieve better performance if add backend pagination with state management.
Maybe, let's see if we can find the cause first. Probably something is being re-rendered over and over.

> Did you try to disable the Store Devtools extension as a workaround?
I get 10 seconds with inspector closed and around 30 with it open.

File: src/app/pages/storage/snapshots/snapshot-table/components/snapshot-details/snapshot-details.component.html [hunk 11]
```diff
@@ -0,0 +1,48 @@
+<ng-container matColumnDef="expandedDetail">
+  <td mat-cell *matCellDef="let snapshot; dataSource: dataSource" [attr.colspan]="colspan">
+    <ix-table-expandable-row [data]="getDetails(snapshot)" [isExpanded]="expandedRow === snapshot">
+      <button
+        mat-button
+        type="button"
+        (click)="doDelete(snapshot)"
+        [attr.tabindex]="expandedRow !== snapshot  ? '-1' : null"
+      >
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/pages/storage/snapshots/snapshot-table/components/snapshot-details/snapshot-details.component.html]: yep, `type` can be omitted and `tabindex` can be removed when `expandable-row` will be hidden from DOM.

**Comment by denysbutenko** [comment]
The PR is updated. I'll resolve unit tests failures soon.

File: src/app/pages/storage/snapshots/snapshot-rollback-dialog/snapshot-rollback-dialog.component.ts (hunk 9 above)
Comment (denysbutenko) [reply src/app/pages/storage/snapshots/snapshot-rollback-dialog/snapshot-rollback-dialog.component.ts:108]: Fixed

File: src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts (hunk 8 above)
Comment (denysbutenko) [reply src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts]: Created a ticket for that - https://jira.ixsystems.com/browse/NAS-114886

**Comment by denysbutenko** [comment]
Ready for another round of review

**Review by undsoft — APPROVED** [review]
Something happens to columns when page is loaded or when snapshot is being deleted. I see weird values in columns for a second.

Otherwise, seems to be working well. I may have missed something as it is a very large PR. Please create multiple PRs in the future. You can also create PRs against an open PR.

File: src/app/pages/account/users/user-list-details/user-list-details.component.spec.ts [hunk 12]
```diff
@@ -33,7 +33,7 @@ describe('UserListDetailsComponent', () => {
     ],
     providers: [
       mockWebsocket([
-        mockCall('user.query', fakeDataSource),
+        mockCall('user.query', fakeUserDataSource),
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/pages/account/users/user-list-details/user-list-details.component.spec.ts:36]: Yeah, it makes sense. This file is a placeholder for tests and has no actual tests. in #6432 will be removed

**Comment by denysbutenko** [comment]
Found an issue, when the master checkbox is still active when the snapshot batch delete dialog is closed after deleting. Fixing...

File: src/app/modules/ix-tables/testing/ix-table.harness.ts [hunk 13]
```diff
@@ -44,6 +44,12 @@ export class IxTableHarness extends MatTableHarness {
     return Promise.resolve(cells);
   }
 
+  async getCellsWithoutExpandedRows(): Promise<string[][]> {
+    const cells = (await this.getCellTextByIndex()).filter((_, index) => index % 2 === 0);
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/modules/ix-tables/testing/ix-table.harness.ts:48]: This is unused method and will be removed

File: src/app/modules/entity/entity-form/validators/has-two-values-validation.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import { ValidatorFn, FormControl } from '@angular/forms';
+
+export function hasTwoValuesValidator(otherControlName: string): ValidatorFn {
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/modules/entity/entity-form/validators/has-two-values-validation.ts]: Added new validator `requiredEmpty` and used with `IxValidatorsService.validateOnCondition`

**Pull request merged**